source/common/http/conn_manager_impl.cc

```diff
@@ -311,8 +313,17 @@ Network::FilterStatus ConnectionManagerImpl::onData(Buffer::Instance& data, bool
 
 void ConnectionManagerImpl::resetAllStreams() {
   while (!streams_.empty()) {
-    // Mimic a downstream reset in this case.
-    streams_.front()->onResetStream(StreamResetReason::ConnectionTermination, absl::string_view());
+    // Mimic a downstream reset in this case. We must also remove callbacks here. Though we are
+    // about to close the connection and will disable further reads, it is possible that flushing
+    // data out can cause stream callbacks to fire (e.g., low watermark callbacks).
+    //
+    // TODO(mattklein123): I tried to actually reset through the codec here, but ran into issues
+    // with nghttp2 state and being unhappy about sending reset frames after the connection had
+    // been terminated via GOAWAY. It might be possible to do something better here inside the h2
+    // codec but there are no easy answers and this seems simpler.
+    auto& stream = *streams_.front();
+    stream.response_encoder_->getStream().removeCallbacks(stream);
```

Comment:
Unfortunately it's more complicated than this. The response encoder itself dies once the stream is gone (either via full decode/encode or reset), so not even a null check here will work. Arguably, we could have an `onStreamDestroyed()` callback where we null out the response_encoder_, but that is a larger change which I would prefer to not conflate with this. In this case, this is definitely safe, since by definition we are cycling through all streams that have not been reset yet (which is further fixed by the other portion of this change). WDYT?

Reply:
I think it's probably fine, but I'm glad this is flagged as high risk :-)